### ISSUE TITLE
listen to push notification

### ISSUE THREAD

**Issue by seme1** [OneSignal/OneSignal-Website-SDK#85]
Is there a way to listen to the push notification ? I may need to perform a certain action when the user receives a certain type of notification while reading a certain article on my website. It would be nice to give users the option of displaying the Chrome notification message _OR_ running a custom script or both. 
The cordova implementation of of notificationOpenedCallback seem to do exactly what I need. However, I am unable to find a similar callback with the Web Push SDK. https://documentation.onesignal.com/docs/phonegap--cordova-sdk-api#notificationOpenedCallback

The only relevant callback for Web Push is addListenerForNotificationOpened but it requires the user to click the displayed message for the callback function to run. Also, it does not give me the option of disabling the default notification message when my website tab is active.

In other words, I may want to tap into the following 
```
self.addEventListener('push', function(event) {  

});

```
as described in the following link:
https://developers.google.com/web/updates/2015/03/push-notifications-on-the-open-web?hl=en

**Comment by jasonpang** [comment]
Hey @seme1,

Unfortunately we don't really have support for running a custom script. Here are some of the limitations (imposed by the browser or by our SDK) and some of the workarounds.

The first issue is that a notification must be displayed whenever a push event is received. From: https://developers.google.com/web/fundamentals/getting-started/push-notifications/step-06?hl=en

> You must pass a `{userVisibleOnly: true}` argument to the `subscribe()` method. This tells the browser that a notification will always be shown when a push message is received. Currently it’s mandatory to show a notification.

Android doesn't have this restriction; this restriction is currently only for web push. So you'll have to display a Chrome notification whenever a push signal is received.

In the next release of the web SDK, we do have a notification displayed event handler (https://github.com/OneSignal/OneSignal-Website-SDK/issues/52). However the notification displayed event fires on the page's JavaScript code, which most likely won't be running once a notification is displayed (before the notification is clicked).

The solution is for our service worker to call some callbacks when a push is received or a notification is displayed to allow you to run some custom code. You would have to modify our service worker to import your script, so that your script would be running in the background even when the site isn't open, just like our service worker does. Unfortunately we don't currently have this extension functionality; we don't currently any callbacks.

If you want to run some general task when a push is received that doesn't depend on the user visiting the page, you can use our Webhooks feature: https://documentation.onesignal.com/docs/webhooks

**Comment by jasonpang** [comment]
User has reached out to us on Intercom's support channel. Closing issue.
